### PR TITLE
feat(mock-server): add all the auth methods

### PR DESCRIPTION
**Pull request by hanspagel**
This PR adds support for *all* auth methods to `@scalar/mock-server`. 🤯 Coming to a `galaxy.scalar.com` near you soon.

It even renders an authorize page on the path set in the OpenAPI document.

**OAuth 2.0 Authorize Page**
Screenshot 2024-10-23 at 15 19 11

**Authentication instructions when booting the Mock Server**
Screenshot 2024-10-23 at 16 46 53

## Release Notes

- **New Features**
  - Introduced OAuth code flow support in the `@scalar/galaxy` package for secure user authentication.
  - Implemented OAuth code flow mocking in the `@scalar/mock-server` package to facilitate testing without actual OAuth interactions.
  - Enhanced OpenAPI specifications with updated security schemes and endpoints for improved API robustness.
  - Added various functions to simulate OAuth 2.0 authorization and token responses.
  - Introduced new utility functions for handling authentication and logging instructions based on security schemes.
  - Added a script for running tests in the `@scalar/galaxy` package.

- **Bug Fixes**
  - Improved error handling for missing specifications and authentication parameters.

- **Tests**
  - Added comprehensive test suites for API key, basic, bearer, and OAuth 2.0 authentication flows to ensure functionality and correctness.
  - Enhanced test coverage for the `createMockServer` function and various authentication scenarios.

### CODE REVIEW

File: packages/mock-server/src/utils/handleAuthentication.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                break
+            }
+          }
+
+          if (isAuthenticated) break
+        }
+        if (isAuthenticated) break
+      }
+
+      if (!isAuthenticated) {
```
[the remark this answers is not in the document]
Comment (hanspagel) [reply packages/mock-server/src/utils/handleAuthentication.ts]: Yes, please make a suggestion!

File: packages/mock-server/src/utils/handleAuthentication.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        for (const [schemeName] of Object.entries(securityRequirement)) {
+          const scheme = schema?.components?.securitySchemes?.[schemeName]
+
+          if (scheme) {
+            switch (scheme.type) {
+              case 'http':
+                if (scheme.scheme === 'basic') {
+                  const authHeader = c.req.header('Authorization')
+
+                  if (authHeader && authHeader.startsWith('Basic ')) {
```
[the remark this answers is not in the document]
Comment (hanspagel) [reply packages/mock-server/src/utils/handleAuthentication.ts:28]: Can youi please add a commitable suggestion?

File: packages/mock-server/src/utils/setupAuthenticationRoutes.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        )
+      }
+
+      if (scheme.flows?.password || scheme.flows?.clientCredentials) {
+        const tokenRoute =
+          (scheme.flows.password?.tokenUrl ||
+            scheme.flows.clientCredentials?.tokenUrl) ??
+          '/oauth/token'
+        app.post(tokenRoute, respondWithToken)
+      }
```
[the remark this answers is not in the document]
Comment (hanspagel) [reply packages/mock-server/src/utils/setupAuthenticationRoutes.ts]: Can you update the committable suggestion?

File: packages/mock-server/src/utils/handleAuthentication.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanspagel) [reply packages/mock-server/src/utils/handleAuthentication.ts]: A lot has changed, can you make an updated suggestion?

**Review by marclave — APPROVED** [review]

**Pull request merged**